pkg/scheduler/scheduler_test.go

```diff
@@ -70,7 +70,7 @@ func Test_afterFunc(t *testing.T) {
 		})
 
 		<-wait
-		assert.InDelta(t, 100*time.Millisecond, end.Sub(start), float64(1*time.Millisecond))
+		assert.InDelta(t, 100*time.Millisecond, end.Sub(start), float64(500*time.Millisecond))
```

Comment:
500ms is a significant variance from the target of 100ms, and instead of the variance being 1%, it's now 500% of the target value 😅

Reply:
Thank you for clarifying.

I think the PR description is pretty clear on why I don't think this test is sound in a fundamental way. Specifically point 3:

> When we write code which uses this scheduler, we can't even safely assume the afterFunc will ever be run, let alone run within a 1ms margin of time error.

Hence the huge allowable delta. No code we write can rely on afterFunc being called within 1ms / 10ms / 50ms of the 100ms deadline because it can't rely on afterFunc being called full stop, and the smaller the allowable delta the more likely this becomes a test of how loaded the test server is.

> With a 500ms variance, if the scheduler were to wait just 10ms, would the test still pass? As 10ms is within 500ms of 100ms

This is an interesting suggestion for a different sanity check. In this case we want the test to assert that `end.Sub(start) >= 100*time.Millisecond`, which seems worthwhile to add and I'll add that.

Reply:
When I wrote this (https://github.com/cert-manager/cert-manager/pull/4233), I totally forgot about the fact that `f` may be called arbitrarily at any moment, and that there is no guarantee that `f` would be called before 100 ms.

Looking at the comment at the start of the file:

```go
// Note that re-implimenting AfterFunc is not a good idea, since testing it
// is tricky as seen in time_test.go in the standard library. We will just
// focus on two important cases: "f" should be run after the duration
```

It seems I did not end this sentence 😅  I think I meant:

```go
// Note that re-implimenting AfterFunc is not a good idea, since testing it
// is tricky as seen in time_test.go in the standard library. We will just
// focus on two important cases: "f" should be run shortly after the duration
// (since cert-manager needs to reliably renew certificates) and afterFunc
// should not leak goroutines.
```

Richard rightly mentioned that [one important piece of comment](https://github.com/cert-manager/cert-manager/pull/4233/files#diff-fb3d6b36d3cd727e1dd2353fb86c0aee0bfaae1c900e0f005ab913620021d551L26-L29) above `afterFunc` was lost when I did the PR:

```go
// For mocking purposes.
// This little bit of wrapping needs to be done because go doesn't do
// covariance, but it does coerce *time.Timer into stoppable implicitly if we
// write it out like so.
var afterFunc = func(c clock.Clock, d time.Duration, f func()) stoppable {
```

I think we still want to make sure `afterFunc` does call `f` within a reasonable delta, since cert-manager users expect the renewal to happen. Unfortunately, we can't run the test for a long time (no one likes long unit tests) which means we have to work around the flakiness due to the short duration of 100 ms.

I propose that we acknowledge the flakiness and that we wrap this test in a `for` loop, similarly as to what the Go library does for [testing afterFunc](https://github.com/golang/go/blob/2580d0e08d5e9f979b943758d3c49877fb2324cb/src/time/sleep_test.go#L239-L243).

Reply:
I have created a follow-up issue https://github.com/cert-manager/cert-manager/issues/4856 with the above suggestion.